akka-http2-support/src/main/scala/akka/http/scaladsl/Http2.scala

```diff
@@ -94,125 +42,105 @@ final class Http2Ext(private val config: Config)(implicit val system: ActorSyste
+    val http1 = Flow[HttpRequest].mapAsync(parallelism)(handler).joinMat(http.serverLayer(settings, None, log))(Keep.left)
+    val http2 = Http2Blueprint.handleWithStreamIdHeader(parallelism)(handler)(system.dispatcher).joinMat(Http2Blueprint.serverStackTls(settings, log))(Keep.left)
+
+    val masterTerminator = new MasterServerTerminator(log)
 
-    createServerRunnableGraph(interface, effectivePort, settings, fullLayer _, log).run()
+    Tcp().bind(interface, effectivePort, settings.backlog, settings.socketOptions, halfClose = false, settings.timeouts.idleTimeout)
+      .mapAsyncUnordered(settings.maxConnections) {
+        incoming: Tcp.IncomingConnection =>
+          try {
+            httpPlusSwitching(http1, http2).addAttributes(Http.prepareAttributes(settings, incoming))
+              .watchTermination()(Keep.right)
+              .joinMat(incoming.flow)(Keep.left)
+              .run().recover {
+                // Ignore incoming errors from the connection as they will cancel the binding.
+                // As far as it is known currently, these errors can only happen if a TCP error bubbles up
+                // from the TCP layer through the HTTP layer to the Http.IncomingConnection.flow.
+                // See https://github.com/akka/akka/issues/17992
+                case NonFatal(ex) =>
+                  Done
+              }(ExecutionContexts.sameThreadExecutionContext)
+          } catch {
+            case NonFatal(e) =>
+              log.error(e, "Could not materialize handling flow for {}", incoming)
+              throw e
+          }
+      }.mapMaterializedValue {
+        _.map(tcpBinding => ServerBinding(tcpBinding.localAddress)(
+          () => tcpBinding.unbind(),
+          timeout => masterTerminator.terminate(timeout)(fm.executionContext)
+        ))(fm.executionContext)
+      }.to(Sink.ignore).run()
   }
 
-  private def createServerRunnableGraph(
-    interface: String, port: Int,
-    settings:    ServerSettings,
-    createLayer: () => Flow[ByteString, ByteString, Future[Done]],
-    log:         LoggingAdapter)(implicit fm: Materializer): RunnableGraph[Future[ServerBinding]] = {
+  type HttpImplementation = Flow[SslTlsInbound, SslTlsOutbound, NotUsed]
+  type HttpPlusSwitching = (HttpImplementation, HttpImplementation) => Flow[ByteString, ByteString, NotUsed]
 
-    val connections = Tcp().bind(interface, port, settings.backlog, settings.socketOptions, halfClose = false, settings.timeouts.idleTimeout)
+  def onlyHttp2(http1: HttpImplementation, http2: HttpImplementation): Flow[ByteString, ByteString, NotUsed] =
+    TLSPlacebo().reversed join http2
 
-    val masterTerminator = new MasterServerTerminator(log)
+  def priorKnowledge(http1: HttpImplementation, http2: HttpImplementation): Flow[ByteString, ByteString, NotUsed] = {
+    def chooseProtocol(sessionBytes: SessionBytes): String =
+      if (sessionBytes.bytes.startsWith(Http2Protocol.ClientConnectionPreface)) "h2" else "http/1.1"
```

Comment:
We only support the case here where the full preface is received in one go (in contrast to the `PriorKnowledgeSwitch`). I'd say it's sensible enough to just require that it's received in one go. Alternatively, we can change `AlpnSwitch` to gather a fixed or configurable number of bytes before making a protocol decision.

Reply:
I think this would be reasonable to support in the future, but that can be a future improvement.